### DISCUSSION TITLE
Add the Waveform as an overlay to the mania editor

### DISCUSSION THREAD

**Discussion by SurfChu85** [ppy/osu#13184] [Ideas]
**Describe the new feature:** 
Aside from the timeline, that shows the waveform above the editor, an overlaid waveform on the editor will complement the default snap lines that are currently being used. This will make identification of the more "obscure" snappings on notecharts/beatmaps that focus more on the technical side of charting/mapping.

**Proposal designs of the feature:** 
An example of this being implemented is on the ArrowVortex chart editor for StepMania
![image](https://user-images.githubusercontent.com/29819386/111459132-41bff080-8755-11eb-9999-ce0c7e4ef123.png)

**Comment by peppy** [comment]
Not sure how useful this will be when you already have snap lines present – even if it shows you detail that you otherwise wouldn't see, you still can't snap to the waveform itself. So unless you're planning to use it as a hint and then trial-and-error switch through the snap divisor until you find the one you are looking for...?

**Comment by SurfChu85** [comment]
> So unless you're planning to use it as a hint and then trial-and-error switch through the snap divisor until you find the one you are looking for...?

Yep, that's what I meant about complementing the snap divisors.

**Comment by sr229** [comment]
I wouldn't exactly like the idea of this kind of design, probably a more greyscale kind of graphing so there's more visibility on things like the snap lines or timing points (if we're going to visualize them in-editor).

**Comment by peppy** [comment]
It's already visible in the timeline for now, so I'm not even sure it's relevant to put it in the playfield as done here. Seems more like a stylistic change than a functional one.

**Comment by Witherbomb-47** [comment]
This would be super helpful for properly timing beatmaps, especially longer ones/those with bpm changes that aren't immediate but happen over time. No more need to listen to a section over and over, questioning if it's synced properly or not, when you can just look at the waveform and the snap overlay.

(necroposting, sorry)

**Reply by peppy** [reply]
There's a timing screen for this?

**Reply by Witherbomb-47** [reply]
Yes, but you can't see how the grid lines up with the audio waveform. With songs that have many bpm changes, or gradual bpm changes, being able to double check what you're hearing with a visual indicator helps a ton. I've messed around with arrow vortex, and their system made it super easy to time stuff properly, it's just that it's a pretty old program that isn't well-maintained.

People have tried using AV to chart stuff instead of osu because of this feature, just because of how helpful it is, but it's difficult to use. Now that it's getting open-sourced, someone might fork it and improve on it and create a better timing program people will use instead of osu.

One thing to note is that I am specifically talking about syncing BPM properly, rather than using the waveform to determine what snapping (1/4, 1/8, etc) should be used. Of course the additional information would probably help with proper snapping and density, the huge change here would be making it much easier to time beatmaps that change bpm often/over time.

**Reply by Witherbomb-47** [reply]
Also: This wouldn't have to be forced on. ArrowVortex has it as a toggleable option, I believe. Osu can just include it as a drop-down option for the people that want to use it (Who're more common than you think!) There should be ample space for it as well, since even 9k only takes up about half the screen in the editor, which should be fine even if both the left and right audio channels are included. Then again, I'm not a UI designer, and this is just my opinion.

**Reply by Witherbomb-47** [reply]
WAIT THIS IS A FEATURE IN LAZER?

I legitimately had no idea, Yeah that explains your confusion, my bad.
